merkle-tree/merkle_tree.go

```diff
@@ -0,0 +1,225 @@
+package merkle
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"hash"
+	"math"
+
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+// A Merkle tree example:
+//
+//              ROOT
+//            /       \
+//          h3         h7
+//         /  \      /    \
+//       h1   h2    h5    h6
+//      / \   / \   / \   / \
+//     L0 L1 L2 L3 L4 L5 L6 L7
+//
+// Each leaf node (L0 - L7) contains a data value, and each intermediate node
+// (h1 - h7) contains the hash of its two child nodes. The root node (ROOT) contains
+// the hash of the entire tree. To prove the inclusion of a leaf node in the tree,
+// a Merkle proof would consist of the leaf node, along with the hash values of its
+// sibling nodes on the path from the leaf to the root.
+//
+// So the proof for leaf node L0 will look like this:
+// H(Hash(L1), Hash(H2), Hash(H7))
+
+var (
+	errLeafNotFound = errors.New("leaf not found")
+)
+
+// MerkleNode represents a single node in merkle tree
+type MerkleNode struct {
+	left   *MerkleNode
+	right  *MerkleNode
+	parent *MerkleNode
+	data   []byte
+	hash   []byte
```

Comment:
Represent it as `types.Hash` perhaps? Eventually if and when this becomes a library, we would have some Hash type alias defined there and use that one instead. WDYT?

Reply:
Why not

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1234/commits/9ce1f4656d13a5c90ec4a241a980c7b0fd433658